core/src/main/java/org/apache/iceberg/AbstractTableScan.java

```diff
@@ -0,0 +1,173 @@
+    Preconditions.checkArgument(
+        tableOps().current().snapshot(scanSnapshotId) != null,
+        "Cannot find snapshot with ID %s",
+        scanSnapshotId);
+    return newRefinedScan(
+        tableOps(), table(), tableSchema(), context().useSnapshotId(scanSnapshotId));
+  }
+
+  public ThisT useRef(String name) {
+    Preconditions.checkArgument(
+        snapshotId() == null, "Cannot override ref, already set snapshot id=%s", snapshotId());
+    Snapshot snapshot = table().snapshot(name);
+    Preconditions.checkArgument(snapshot != null, "Cannot find ref %s", name);
+    return newRefinedScan(
+        tableOps(), table(), tableSchema(), context().useSnapshotId(snapshot.snapshotId()));
+  }
+
+  public ThisT asOfTime(long timestampMillis) {
+    Preconditions.checkArgument(
+        snapshotId() == null, "Cannot override snapshot, already set snapshot id=%s", snapshotId());
+
+    return useSnapshot(SnapshotUtil.snapshotIdAsOfTime(table(), timestampMillis));
+  }
+
+  @Override
+  public CloseableIterable<T> planFiles() {
+    Snapshot snapshot = snapshot();
+    if (snapshot != null) {
+      LOG.info(
+          "Scanning table {} snapshot {} created at {} with filter {}",
+          table(),
+          snapshot.snapshotId(),
+          DateTimeUtil.formatTimestampMillis(snapshot.timestampMillis()),
+          ExpressionUtil.toSanitizedString(filter()));
+
+      Listeners.notifyAll(new ScanEvent(table().name(), snapshot.snapshotId(), filter(), schema()));
+      List<Integer> projectedFieldIds = Lists.newArrayList(TypeUtil.getProjectedIds(schema()));
+      List<String> projectedFieldNames =
+          projectedFieldIds.stream().map(schema()::findColumnName).collect(Collectors.toList());
+
+      Timer.Timed planningDuration = scanMetrics().totalPlanningDuration().start();
+
+      return CloseableIterable.whenComplete(
+          doPlanFiles(),
+          () -> {
+            planningDuration.stop();
+            Map<String, String> metadata = Maps.newHashMap(context().options());
+            metadata.putAll(EnvironmentContext.get());
+            ScanReport scanReport =
+                ImmutableScanReport.builder()
+                    .schemaId(schema().schemaId())
+                    .projectedFieldIds(projectedFieldIds)
+                    .projectedFieldNames(projectedFieldNames)
+                    .tableName(table().name())
+                    .snapshotId(snapshot.snapshotId())
+                    .filter(ExpressionUtil.sanitize(filter()))
+                    .scanMetrics(ScanMetricsResult.fromScanMetrics(scanMetrics()))
+                    .metadata(metadata)
+                    .build();
+            context().metricsReporter().report(scanReport);
+          });
+    } else {
+      LOG.info("Scanning empty table {}", table());
+      return CloseableIterable.empty();
+    }
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public CloseableIterable<C> planTasks() {
```

Comment:
I don't think this cast is generally safe as we don't know the boundary and `planTaskGroups` would always create a specific type of a task group. I'd remove this from here and implement this method in the position deletes table scan itself. If we have more use cases later, we can add a common class for all batch scans, just like we have one for all table scans now. That way, you also skip the cast completely.

Reply:
Done